internal/langserver/handlers/command/module_providers.go

```diff
@@ -0,0 +1,85 @@
+package command
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/creachadair/jrpc2/code"
+	lsctx "github.com/hashicorp/terraform-ls/internal/context"
+	"github.com/hashicorp/terraform-ls/internal/langserver/cmd"
+	op "github.com/hashicorp/terraform-ls/internal/terraform/module/operation"
+	"github.com/hashicorp/terraform-ls/internal/uri"
+	tfaddr "github.com/hashicorp/terraform-registry-address"
+)
+
+const moduleProvidersVersion = 0
+
+type moduleProvidersResponse struct {
+	FormatVersion        int                            `json:"v"`
+	ProviderRequirements map[string]providerRequirement `json:"provider_requirements"`
+	InstalledProviders   map[string]string              `json:"installed_providers"`
+}
+
+type providerRequirement struct {
+	DisplayName       string `json:"display_name"`
+	VersionConstraint string `json:"version_constraint,omitempty"`
+	DocsLink          string `json:"docs_link,omitempty"`
+}
+
+func ModuleProvidersHandler(ctx context.Context, args cmd.CommandArgs) (interface{}, error) {
+	response := moduleProvidersResponse{
+		FormatVersion:        moduleProvidersVersion,
+		ProviderRequirements: make(map[string]providerRequirement),
+		InstalledProviders:   make(map[string]string),
+	}
+
+	modUri, ok := args.GetString("uri")
+	if !ok || modUri == "" {
+		return response, fmt.Errorf("%w: expected module uri argument to be set", code.InvalidParams.Err())
+	}
+
+	if !uri.IsURIValid(modUri) {
+		return response, fmt.Errorf("URI %q is not valid", modUri)
+	}
+
+	modPath, err := uri.PathFromURI(modUri)
+	if err != nil {
+		return response, err
+	}
+
+	mm, err := lsctx.ModuleFinder(ctx)
+	if err != nil {
+		return response, err
+	}
+
+	mod, _ := mm.ModuleByPath(modPath)
+	if mod == nil {
+		return response, nil
+	}
+
+	if mod.MetaState == op.OpStateUnknown || mod.MetaErr != nil {
+		return response, nil
+	}
+
+	for provider, version := range mod.Meta.ProviderRequirements {
+		response.ProviderRequirements[provider.String()] = providerRequirement{
+			DisplayName:       provider.ForDisplay(),
+			VersionConstraint: version.String(),
+			DocsLink:          getProviderDocumentationLink(provider),
+		}
+	}
+
+	for provider, version := range mod.InstalledProviders {
+		response.InstalledProviders[provider.String()] = version.String()
+	}
+
+	return response, nil
+}
+
+func getProviderDocumentationLink(provider tfaddr.Provider) string {
+	if provider.IsLegacy() || provider.IsBuiltIn() {
```

Comment:
```suggestion
	if provider.IsLegacy() || provider.IsBuiltIn() || provider.Hostname != "registry.terraform.io" {
```

We cannot tell where the docs are if the provider isn't hosted by the official Registry, so we should also exclude linking to these. Such providers can either be side-loaded or hosted in a private registry.

btw. we make the same decision in https://github.com/hashicorp/terraform-schema/blob/188460bcdf1add8d8d7d06fe7aa05ff2cb035a0d/schema/convert_json.go#L349-L366 when attaching document links to the schema for providers.

Reply:
Good catch.

Would it make sense to move `providerHasDocs` at some point into the [terraform-registry-address](https://github.com/hashicorp/terraform-registry-address/blob/main/provider.go) package? Since we're starting to do the same check in different other packages?

Reply:
I'm not sure as I don't have much visibility into the plans around private registry myself, I think the duplication is not necessarily bad in this case. I reckon private registry will eventually have docs too, but I have no idea how that would work in terms of URLs. I reckon some of these assumptions may be challenged as part of https://github.com/hashicorp/vscode-terraform/issues/737 and https://github.com/hashicorp/terraform-ls/issues/458 so I would be hesitant to bake them into a public API of some kind.

Similarly the builtin provider actually does have docs, it's just that we don't have _versioned_ docs for TF Core and we preferred to not make a blind assumption by linking to https://www.terraform.io/docs/language/state/remote-state-data.html

So many things are still "WIP", unfortunately.